services/dis-promrulegroups-operator/internal/controller/prometheusrule_controller.go

```diff
@@ -0,0 +1,450 @@
+		// The names are the same for every cluster so we need to prefix them
+		if !strings.Contains(promRuleCopy.Spec.Groups[idx].Name, r.AzClusterName) {
+			promRuleCopy.Spec.Groups[idx].Name = r.AzClusterName + "-" + ruleGroup.Name
+		}
+		interval, err := prometheusmodel.ParseDuration(string(*ruleGroup.Interval))
+		if err != nil {
+			log.Error(err, "Failed to parse the Interval from the PrometheusRule Spec")
+			return "", err
+		}
+		// Can't be lower than 1m.
+		if interval < prometheusmodel.Duration(1*time.Minute) {
+			*ruleGroup.Interval = monitoringv1.Duration("1m")
+		}
+	}
+
+	marshalledPromRule, err := json.Marshal(promRuleCopy.Spec)
+
+	if err != nil {
+		log.Error(err, "Failed to marshal the promRule")
+		return "", err
+	}
+
+	tmp := strings.Split(r.AzAzureMonitorWorkspace, "/")
+	azureMonitorWorkspaceName := tmp[len(tmp)-1]
+
+	cmd := exec.Command(
+		r.NodePath,
+		r.AzPromRulesConverterPath,
+		"--azure-monitor-workspace",
+		azureMonitorWorkspaceName,
+		"--location",
+		r.AzResourceGroupLocation,
+		"--cluster-name",
+		r.AzClusterName,
+	)
+
+	var in bytes.Buffer
+	var out, errb strings.Builder
+
+	in.Write([]byte(marshalledPromRule))
+
+	cmd.Stdin = &in
+	cmd.Stdout = &out
+	cmd.Stderr = &errb
+	err = cmd.Run()
+	if err != nil {
+		log.Error(err, "Failed to convert PrometheusRule into PrometheusRuleGroup", "Stderr", errb.String())
+		return "", err
+	}
+	jsonString := out.String()
+
+	return jsonString, nil
+}
+
+func timestamp() string {
+	now := time.Now()
+
+	var sb strings.Builder
+	sb.WriteString(strconv.Itoa(now.Year()))
+	sb.WriteString(strconv.Itoa(int(now.Month())))
+	sb.WriteString(strconv.Itoa(now.Day()))
+	sb.WriteString(strconv.Itoa(now.Hour()))
+	sb.WriteString(strconv.Itoa(now.Minute()))
+	sb.WriteString(strconv.Itoa(now.Second()))
+
+	return sb.String()
+}
+
+func hasAllAnnotations(annotations map[string]string) bool {
+	boolRes := true
+	for _, a := range allAnnotations {
+		_, ok := annotations[a]
+		boolRes = boolRes && ok
+	}
+	return boolRes
+}
+
+func generateArmDeploymentName(req ctrl.Request, suffix string) string {
+	// Limit of 64 characters
+	return req.Namespace + "-" + req.Name + "-" + suffix
+}
+func removedGroups(old, new []string) []string {
+	groupsToRemove := make([]string, 0)
+	for _, viOld := range old {
+		found := false
+		for _, viNew := range new {
+			if viNew == viOld {
+				found = true
+				break
+			}
+		}
+		if !found {
+			groupsToRemove = append(groupsToRemove, viOld)
+		}
+	}
+	return groupsToRemove
+}
+
+func generateRuleGroupNamesAnnotationString(promRule monitoringv1.PrometheusRule) string {
+	resourceNames := ""
+	for idx, p := range promRule.Spec.Groups {
+		if idx+1 < len(promRule.Spec.Groups) {
+			resourceNames = resourceNames + p.Name + ","
+		} else {
+			resourceNames = resourceNames + p.Name
+		}
+	}
+	return resourceNames
+}
+
+func hashArmTemplate(bytes []byte) string {
+	h := sha256.New()
+	h.Write(bytes)
+	return base64.URLEncoding.EncodeToString(h.Sum(nil))
```

Comment:
Probably not strictly necessary but makes it more legible. The annotation would have some weird characters otherwise.
Also, the [docs](https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/) do talk about having: **_The keys and the values in the map must be strings. In other words, you cannot use numeric, boolean, list or other types for either the keys or the values._** so I'd rather not find myself having weird bugs later on. I suppose I could have also just wrapped it in a string() call